engines/dfc_provider/app/services/supplied_product_builder.rb

```diff
@@ -36,6 +44,23 @@ def self.import_variant(supplied_product)
     end
   end
 
+  def self.referenced_spree_product(supplied_product)
+    uri = supplied_product.spree_product_uri
+    id = supplied_product.spree_product_id
+
+    if uri.present?
+      route = Rails.application.routes.recognize_path(uri)
+      params = Rack::Utils.parse_nested_query(URI.parse(uri).query)
+
+      # Check that the given URI points to us:
+      return unless uri == urls.enterprise_url(route.merge(params))
+
+      Spree::Product.find(params["spree_product_id"])
+    elsif id.present?
+      Spree::Product.find(id)
```

Comment:
I think that the product search should be scoped to `current_enterprise.supplied_products.find_by(id: ...)`. And if we don't find it, we just ignore it. That will check the permission and also allow someone to clone the products of another enterprise when they have only read permissions. 

The controller needs to pass the enterprise for this. We can then also assign the enterprise here instead of in the controller.